blaze-client/src/main/scala/org/http4s/client/blaze/PoolManager.scala

```diff
@@ -1,66 +1,146 @@
-private final class PoolManager (maxPooledConnections: Int, builder: ConnectionBuilder) extends ConnectionManager {
+import scalaz.{-\/, \/-, \/}
+import scalaz.syntax.either._
+import scalaz.concurrent.{Actor, Task}
 
-  require(maxPooledConnections > 0, "Must have finite connection pool size")
+private object PoolManager {
+  type Callback[A] = Throwable \/ A => Unit
+  sealed trait Protocol
+  case class RequestConnection(key: RequestKey, callback: Callback[BlazeClientStage]) extends Protocol
+  case class ReturnConnection(key: RequestKey, stage: BlazeClientStage) extends Protocol
+  case class DisposeConnection(key: RequestKey) extends Protocol
+  case class Shutdown(callback: Callback[Unit]) extends Protocol
+}
+import PoolManager._
 
-  private case class Connection(scheme: Option[Scheme], auth: Option[Authority], stage: BlazeClientStage)
+private final class PoolManager(builder: ConnectionBuilder,
+                                maxTotal: Int)
+  extends ConnectionManager {
 
   private[this] val logger = getLogger
-  private var closed = false  // All access in synchronized blocks, no need to be volatile
-  private val cs = new mutable.Queue[Connection]()
+  private var isClosed = false
+  private var allocated = 0
+  private val idleQueue = new mutable.Queue[BlazeClientStage]
+  private val waitQueue = new mutable.Queue[Callback[BlazeClientStage]]()
 
-  /** Shutdown this client, closing any open connections and freeing resources */
-  override def shutdown(): Task[Unit] = Task.delay {
-    logger.debug(s"Shutting down ${getClass.getName}.")
-    cs.synchronized {
-      if(!closed) {
-        closed = true
-        cs.foreach(_.stage.shutdown())
+  private def stats =
+    s"allocated=${allocated} idleQueue.size=${idleQueue.size} waitQueue.size=${waitQueue.size}"
+
+  private def createConnection(key: RequestKey): Unit = {
+    if (allocated < maxTotal) {
+      allocated += 1
+      logger.debug(s"Creating connection: ${stats}")
+      Task.fork(builder(key)).runAsync {
+        case \/-(stage) =>
+          logger.debug(s"Submitting fresh connection to pool: ${stats}")
+          actor ! ReturnConnection(key, stage)
+        case -\/(t) =>
+          logger.error(t)("Error establishing client connection")
+          actor ! DisposeConnection(key)
       }
     }
+    else {
+      logger.debug(s"Too many connections open.  Can't create a connection: ${stats}")
+    }
   }
 
-  override def recycleClient(request: Request, stage: BlazeClientStage): Unit =
-    cs.synchronized {
-      if (closed) stage.shutdown()
-      else {
-          logger.debug("Recycling connection.")
-          cs += Connection(request.uri.scheme, request.uri.authority, stage)
+  private val actor = Actor[Protocol] {
+    case RequestConnection(key, callback) =>
+      logger.debug(s"Requesting connection: ${stats}")
+      if (!isClosed) {
+        def go(): Unit = {
+          if (idleQueue.nonEmpty) {
+            idleQueue.dequeue() match {
```

Comment:
Am I missing something or does the `key` go unused? It seems to me that the queue just gives the first connection regardless of where it may be connected to.